### PR TITLE
nixosTests.nano: replace with script using GNU expect

### PR DESCRIPTION
**Pull request by Synthetica9**
###### Motivation for this change

cc @NeQuissimus 

There was no real reason to have a NixOS test for nano, might as well use the right tool for the job. Not sure if the style I used here (seperate file with `callPackage`) is encouraged, would like some feedback on that. Should be a lot quicker, easier to run, and work on MacOS as well. 

If this is merged I'd love to write some other tests for other popular TUI packages (htop and such)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by risicle** [comment]
On macos 10.15
```
building '/nix/store/9wh82d4kmxz3x1kfvimkrfr69c302vxh-nano-test-expect.drv'...
/nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 4: spawn: command not found
/nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 5: expect: command not found
/nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 8: send: command not found
/nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 9: expect: command not found
/nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 12: send: command not found
/nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 13: expect: command not found
/nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 16: send: command not found
/nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 17: expect: command not found
/nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 20: send: command not found
grep: /tmp/file.txt: No such file or directory
```

File: pkgs/applications/editors/nano/test-with-expect.nix [hunk 1]
```diff
@@ -0,0 +1,34 @@
+{ nano, expect, runCommand, writeScriptBin }:
+
+let expectScript = writeScriptBin "expect-script" ''
+  #!${expect}/bin/expect -f
+
+  # Load nano
+  spawn nano /tmp/file.txt
```
Comment (risicle) [inline pkgs/applications/editors/nano/test-with-expect.nix:7]: You should use `$TMPDIR` (or even `mktemp -d`) instead of expecting `/tmp` to exist.

Comment (Synthetica9) [reply pkgs/applications/editors/nano/test-with-expect.nix]: eh I guess I'll just use cwd, thanks for pointing it out.

Comment (Synthetica9) [reply pkgs/applications/editors/nano/test-with-expect.nix]: eh I guess I'll just use cwd, thanks for pointing it out.

**Comment by Synthetica9** [comment]
> On macos 10.15
> 
> ```
> building '/nix/store/9wh82d4kmxz3x1kfvimkrfr69c302vxh-nano-test-expect.drv'...
> /nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 4: spawn: command not found
> /nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 5: expect: command not found
> /nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 8: send: command not found
> /nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 9: expect: command not found
> /nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 12: send: command not found
> /nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 13: expect: command not found
> /nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 16: send: command not found
> /nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 17: expect: command not found
> /nix/store/vg05ysldddwcfxsp0yiw8xzsi4f0kfm4-expect-script/bin/expect-script: line 20: send: command not found
> grep: /tmp/file.txt: No such file or directory
> ```

wait... what's happening here? Does macos not support shebangs?

**Comment by risicle** [comment]
https://stackoverflow.com/questions/67100831/macos-shebang-with-absolute-path-not-working/67101108#67101108

> MacOS doesn't support shebangs pointing at scripts -- this causes an Exec format error...

**Comment by Synthetica9** [comment]
> [stackoverflow.com/questions/67100831/macos-shebang-with-absolute-path-not-working/67101108#67101108](https://stackoverflow.com/questions/67100831/macos-shebang-with-absolute-path-not-working/67101108#67101108)
> 
> > MacOS doesn't support shebangs pointing at scripts -- this causes an Exec format error...

I suppose I'll just sidestep the shebang issue by calling `expect` directly. Would you care to try again?

File: pkgs/applications/editors/nano/test-with-expect.nix [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+  # Send "return" to accept the file path.
+  send "\r"
+  sleep 1
+  exit
+''; in
+runCommand "nano-test-expect"
+{
+  nativeBuildInputs = [ nano expect ];
+  passthru = { inherit expect-script; };
```
Comment (risicle) [inline pkgs/applications/editors/nano/test-with-expect.nix:30]: Is this passthru needed?

**Review by risicle — APPROVED** [review]
WFM, nixos x86_64 & macos 10.15

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 161707` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Comment by andersk** [comment]
This broke the nixos:trunk-combined:tested job (or at least contributed to breaking it: #162317) at

https://github.com/NixOS/nixpkgs/blob/8725b29aa0ff69753d2daef5beb6abb454d29cfc/nixos/release-combined.nix#L101

18bd82edcc752d6a0e6cce1401ba0c81353a03ca removes that line. Since there doesn’t seem to be a way for Hydra aggregate jobs to refer to `passthru.tests` (due to the stripping performed by [`hydraJob`](https://github.com/NixOS/nixpkgs/blob/master/lib/customisation.nix#L171)), we’ll need to keep this in mind when converting some tests to this style.

**Pull request merged**